onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc

```diff
@@ -89,39 +118,64 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputP
 
   Initializer unpacked_tensor(*new_shape_tensor);
   auto new_shape = unpacked_tensor.DataAsSpan<int64_t>();
+
   if (new_shape.empty()) {
     LOGS(logger, VERBOSE) << "New shape of reshape cannot be empty";
     return false;
   }
 
+  // CoreML reshape does not support 0 as a dimension
+  // Even though MLProgram Reshape docs imply 0 is supported as a dimension, it is not supported
+  if (allow_zero) {
+    if (std::find(new_shape.begin(), new_shape.end(), int64_t{0}) != new_shape.end()) {
+      LOGS(logger, VERBOSE) << "Reshape does not support new shape with 0. "
+                            << "New shape: " << Shape2String(new_shape);
+      return false;
+    }
+  }
+
   std::vector<int64_t> input_shape;
-  if (!GetStaticShape(*input_defs[0], input_shape, logger))
-    return false;
 
-  if (input_shape.empty()) {
-    LOGS(logger, VERBOSE) << "Reshape does not support empty input shape";
+  // first input must be fixed rank OR (first input has variadic rank AND shape only contains positive integers)
+  // as per docs, 0 is considered an illegal shape element if the input is variadic
+  if (!GetShape(*input_defs[0], input_shape, logger)) {
+    if (!AllPositiveShape(new_shape)) {
+      LOGS(logger, VERBOSE) << "Unable to get shape of input -- input must have fixed rank for reshape, unless the new shape contains all positive integers. "
+                               "Input shape: "
+                            << Shape2String(input_shape) << ", new shape: " << Shape2String(new_shape);
+      return false;
+    } else {
+      return true;
+    }
```

Comment:
We return true if the input shape is unknown and AllPositiveShape is true, but if the input shape is known we have additional checks like `new_shape.size() > 5`. should we be falling through to at least some of the later checks instead of returning true here?

Reply:
Ah this is a good point but ended up removing support for this case because there's no good way to write unit tests for it at the moment and only affects 1 node in the ChromaDB model -- will keep this in mind for future implementation though